compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -880,38 +888,48 @@ fn transform_ty<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>, options: TransformTyOptio
                     // Generalize any repr(transparent) user-defined type that is either a pointer
                     // or reference, and either references itself or any other type that contains or
                     // references itself, to avoid a reference cycle.
+
+                    // If the self reference is not through a pointer, for example, due
+                    // to using `PhantomData`, need to skip normalizing it if we hit it again.
+                    let mut parents = parents.to_vec();
+                    parents.push(ty);
```

Comment:
so... given this is recursive, this is just gonna be unbounded memory usage, huh...?

I cannot agree we should merge this without someone answering first how much recursion it takes to kill the compiler here?

Reply:
So, the size of `parents` only increases when it unwraps a transparent type it hasn't already visited. This means the size of `parents` is limited to a spine of `#[repr(transparent)]` types in depth.

I currently have it optimized to avoid any allocations in the common case (no transparents) by taking a `&[]`. If we want to make it work in a way that's optimal for maximal memory usage, I can make it take a `&mut Vec<u8>` instead, and `.truncate()` after this if statement to reset it. (That would make the total memory used by this O(n) instead of O(n^2)).

I just went to double check, and evidently `Vec::new()` doesn't cost an allocation if you don't ever push to it. Would you like me to switch to that?

Reply:
To be clear, I may or may not decide the amount of recursion this handles is acceptable, but there's a number and I want it.

Reply:
Do you mean that you want the stack depth that the compiler goes to before dying without this patch, or do you mean you want to know how deeply I can nest transparent structs before the compiler gives up? (I'm guessing they'll be similar, because it'll be based on the stack frame count, not this vector, but I just want to make sure I understand you correctly.)

Reply:
how deep the struct nesting must be, yes.

Reply:
If it dies because of sigsegv instead, that's reassuring!

Reply:
Seems like it needs to be [pretty deep](https://gist.github.com/maurer/13d0e895525a4f6ee3477970df21dd77) before it dies. This will stop dying with one layer of alternation removed in today's nightly `rustc`, and will stop dying with two layers of alternation removed with the parent tracking (slightly bigger stack frame, so that's expected).

Reply:
nice.